decisions/2020/06/25-continuous-app-delivery-discussion.md

```diff
@@ -0,0 +1,81 @@
+# 2020-06-25, Continuous App Delivery
+
+## Attending
+
+Austin, Erik, Lars, Stian, Zubair, Birk
+
+## Context
+
+The meeting was scheduled to identify and propose solutions for features that we need to support continuous app delivery for core applications.  
+
+Link to related issue and agenda: https://github.com/dhis2/notes/issues/114
+
+
+## Support multiple versions of the same app
+
+- This was top priority before the meeting, as we wanted to have admins test apps on the instance before installing a new versions.
+- People are accustomed to have a staging server, so the need for this is not as urgent.
+
+**Conclusion:**
+
+- This has been postponed for now as there are more urgent issues.
+- A nice to have down the line.
+
+
+## Support for overriding core apps by custom apps
+
+- To be able to update core apps we need them to install them as a "custom app".
+- The proposed solution was to have the server check for a custom app of the same name, and use that if it exists.
+- We need to support "old" urls (eg. "dhis-web-dashboard") for backwards compatability, these should redirect to a possibly new url (eg. "/apps")
+- The issue of collisions of app names were brought up, as users could unintentionally install apps by the same name of the core app, thus overriding it.
+- Installing apps should be a locked down permission, and admins should be aware of conflicts.
+- We could potentially notify the user the first time a "custom core app" is installed.
+
+**Conclusion:**
+
+- Going forward with implementation of overrides of core apps.
+- Jira issue for this: https://jira.dhis2.org/browse/DHIS2-4394
+
+## Re-introduce the proxy through core to App Hub
+
+- The proxy was removed mostly due to the ease of introducing new APIs between the App Hub and management app. 
+- Concerns were raised of the removal of the proxy to App Hub.
+    - Air gaps, as clients may not have internet access or have poor internet connections.
+- Security needs to be considered
+- When reintroducing, we need to implement new endpoints like "/channels", and verify that all endpoints are working. 
```

Comment:
Instead of implementing this as a 1:1 API Proxy, could we consider adding a dynamic pass-through proxy to the App Hub through the Core so if we do not need to change the Core API end-points when adding/changing App Hub API end-points?

Reply:
I totally agree. I also brought up having a dynamic pass-through but Lars didn't seem to approve of the idea due to security issues. However, if just the App Hub URL is in the allowlist, I don't really see an issue what that.

pinging @larshelge @amcgee

Reply:
We should hook in our security dude on that discussion as well if there are safe enough ways to accomplish what we need (not duplicating the entire the App Hub API to the DHIS2 Core API).

That was "OK" before because there was one end-point that the Core API needed (get all apps), but the situation going forward is different and changes rapidly so we might need to implement an API in the App Hub API, and then duplicate it in the Core, and then implement it in the App Management App and wait 6 months before it is rolled out.

One of the priorities of not using a proxy was to allow older instances of DHIS2 to utilize new features on the App Hub if they only update their App Management App.

Reply:
> I also brought up having a dynamic pass-through but Lars didn't seem to approve of the idea due to security issues. However, if just the App Hub URL is in the allowlist, I don't really see an issue what that, maybe I'm missing something?

I don't see the problem here either, and there are only mentions of generic "security issues" in these notes and the JIRA issues, so could we please note exactly what security issues we are talking about to see if we can mitigate them?

Reply:
@larshelge said that Morten S (security engineer) would likely be the one to investigate application management security (namely HTTPS certificate validation and the possibility of hash validation).  It would be worth investigating the vulnerability surface area of a wildcard proxy.

A design requirement is definitely to avoid any need for custom WAR files, and a design goal is to support optional distribution as JAR files with no bundled apps (eventually, not initially)

Reply:
Thanks for the additional context. Would be good to include those points in the conclusion and/or description paragraph.

Reply:
+1 agreed!  @Birkbjo